src/main/resources/XWiki/Macros/Team.xml

```diff
@@ -0,0 +1,1505 @@
+        <multiSelect>0</multiSelect>
+        <name>use</name>
+        <number>3</number>
+        <prettyName>Use this extension</prettyName>
+        <relationalStorage>0</relationalStorage>
+        <separator> </separator>
+        <separators>|, </separators>
+        <size>1</size>
+        <unmodifiable>0</unmodifiable>
+        <values>currentPage|onDemand|always</values>
+        <classType>com.xpn.xwiki.objects.classes.StaticListClass</classType>
+      </use>
+    </class>
+    <property>
+      <cache>long</cache>
+    </property>
+    <property>
+      <code>(function () {
+  function filter (teamDiv) {
+    const l = this.value.toLowerCase();
+    const users = teamDiv.getElementsByClassName('xwikiteam-user');
+    for (let i = 0, len = users.length; i &lt; len; i++) {
+      const user = users[i];
+      user.style.display = (
+        user.textContent.toLowerCase().includes(l) ||
+        user.dataset.username.toLowerCase().includes(l) ||
+        user.dataset.nameWithoutAccents.includes(l)
+      ) ? "" : "none";
+    }
+  }
+
+  function applyCheckbox(teamDiv, checkbox) {
+    if (checkbox.checked) {
+      teamDiv.classList.remove('xwikiteam-usernames-hidden');
+    } else {
+      teamDiv.classList.add('xwikiteam-usernames-hidden');
+    }
+  }
+
+  function bindShowUsernameCheckbox(teamDiv, checkbox) {
+    applyCheckbox(teamDiv, checkbox);
+    checkbox.onclick = applyCheckbox.bind(null, teamDiv, checkbox);
+  }
+
+  function toggleTools(teamDiv, e) {
+    const tools = teamDiv.querySelector('.xwikiteam-tools');
+    if (tools) {
+      tools.hidden = !tools.hidden;
+      if (!tools.hidden) {
+        const filterInput = tools.querySelector('.xwikiteam-filter');
+        if (filterInput) {
+          filterInput.focus();
+        }
+      }
+    }
+    if (e) {
+      e.preventDefault();
+    }
+  }
+
+  function init() {
+    const teamDivs = document.getElementsByClassName('xwikiteam');
+    for (let i = 0, len = teamDivs.length; i &lt; len; i++) {
+      const teamDiv = teamDivs[i];
+      const a = teamDiv.querySelector('.xwikiteam-tools-link a');
+      if (a) {
+        if (a.onclick) {
+          continue; // don't initialize already initialized team macros
+        }
+        a.onclick = toggleTools.bind(null, teamDiv);
+        a.parentNode.hidden = false;
+      }
+      const ul = teamDiv.querySelector('.xwikiteam-ul');
+      const tools = teamDiv.querySelector('.xwikiteam-tools');
+      if (!tools) continue;
+      const filterInput = tools.querySelector('.xwikiteam-filter');
+      if (filterInput) {
+        filterInput.addEventListener('keyup', filter.bind(filterInput, teamDiv));
+        if (filterInput.value) {
+          filter.call(filterInput, teamDiv);
+          toggleTools(teamDiv)
+        }
+      }
+
+      bindShowUsernameCheckbox(teamDiv, teamDiv.querySelector(".xwikiteam-show-username-checkbox"));
+    }
+  }
+
+  document.addEventListener('xwiki:dom:updated', init);
```

Comment:
So, it seems like XWiki indeed triggers `xwiki:dom:updated` either using JQuery, either `document.fire`, which is a [Prototype JS function](https://github.com/prototypejs/prototype/blob/master/src/prototype/dom/event.js#L845).

`document.fire` seems to use [document.createEvent](https://developer.mozilla.org/en-US/docs/Web/API/Document/createEvent), [Event.initEvent](https://developer.mozilla.org/en-US/docs/Web/API/Event/initEvent) and [EventTarget.dispatchEvent](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent), so that should work.

However, it seems it does not work if JQuery is used to trigger the event: only one alert is show with the following code:

```html
<!DOCTYPE html>
<script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
<script>
  addEventListener("DOMContentLoaded", function () {
    document.addEventListener("testcustomevent", function () {
      alert("triggered natively");
    })

    $(document).on("testcustomevent", function () {
      alert("triggered from JQuery");
    })

    $(document).trigger("testcustomevent", {})
  })
</script>
```
Anyway, the [XWiki documentation](https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/FrontendResources/JavaScriptAPI/) advises to use `document.observe`

Now, I can notice that the xwiki:dom:updated event is indeed triggered and that the initTools function from my last commit is called. However, when editing the macro with the WYSIWYG editor, I noticed that the script does not find the `xwikiteam` elements, and that seems to be because the script is run in the context of the outer window, not from the iframe editor.

Is this a bug or a feature (of the platform)?

Anyway, for this macro, it's fine for me that the tools don't appear in the editor (for now) and it would be fine if they would at some point. The tools are intended for viewing. They seem rather useless for edition but they would not hurt neither.

Reply:
> Anyway, the [XWiki documentation](https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/FrontendResources/JavaScriptAPI/) advises to use document.observe

That part of the documentation is outdated. You shouldn't use Prototype.js anymore. You should use jQuery for listening to **custom** XWiki events such as ``xwiki:dom:updated``. For standard events plain JavaScript is perfectly fine.

> However, when editing the macro with the WYSIWYG editor, I noticed that the script does not find the xwikiteam elements, and that seems to be because the script is run in the context of the outer window, not from the iframe editor.

The JavaScript code is executed in the context of the Window that loads it:
* for in-place WYSIWYG editing there's no inner frame / window, so all the JavaScript code is loaded in the main / top level window, and it should be able to find the ``xwikiteam`` elements within the edited content
* for stand-alone WYSIWYG editing the edited content is within an iframe, so nested Window; in this case it depends who loads the JavaScript code; if it's the edited content, then the JavaScript code should be loaded inside the nested Window. But not that JavaScript execution is "disabled" by default in the nested content Window by default, see https://extensions.xwiki.org/xwiki/bin/view/Extension/CKEditor+Integration#HExecuteJavaScriptcodeinsidetheeditingarea28XWiki10.102B29 .

Reply:
Ok, thanks for the information.